src/renderer/store/selectors/contacts.ts

```diff
@@ -183,79 +185,157 @@ export interface IDirectMessage {
   blockedUsers: string[]
 }
 
-export const directMessages = address =>
-  createSelector(
-    messages(address),
-    channelOwner(address),
-    (messages, channelOwner: string): IDirectMessage => {
-      const channelModerators = []
-      const messsagesToRemove: DisplayableMessage[] = []
-      const blockedUsers = []
-      let visibleMessages: DisplayableMessage[] = []
-      for (const msg of messages.reverse()) {
-        switch (msg.type) {
-          case MessageType.AD:
-            if (!blockedUsers.includes(msg.publicKey)) {
-              visibleMessages.push(msg)
-            }
-            break
-          case MessageType.BASIC:
-            if (!blockedUsers.includes(msg.publicKey)) {
-              visibleMessages.push(msg)
+const currentChannel = createSelector(
+  contacts,
+  channelSelector.address,
+  channelSelector.id,
+  (contacts, address, id) => {
+    if (contacts[address]) {
+      return contacts[address]
+    } else {
+      return contacts[id]
+    }
+  }
+)
+
+const usersCertificateMapping = createSelector(
+  certificatesSelector.usersCertificates,
+  (certificates) => {
+    return certificates.reduce<{ [pubKey: string]: { username: string; onionAddress: string; peerId: string } }>((acc, current) => {
+      let parsedCerficated
+      let certObject
+      let nickname = null
+      let onionAddress = null
+      let peerId = null
+      if (current !== null && current) {
+        parsedCerficated = extractPubKeyString(current)
+        certObject = loadCertificate(current)
+        if (certObject.subject.typesAndValues.length === 3) {
+          nickname = certObject.subject.typesAndValues[0].value.valueBlock.value
+          onionAddress = certObject.subject.typesAndValues[1].value.valueBlock.value
+          peerId = certObject.subject.typesAndValues[2].value.valueBlock.value
+        } else {
+          return
+        }
+      }
+      acc[parsedCerficated] = {
+        username: nickname,
+        onionAddress: onionAddress,
+        peerId: peerId
+      }
+      return acc
+    }, {})
+  }
+)
+
+const messagesOfChannelWithUserInfo = createSelector(
+  currentChannel, usersCertificateMapping,
+  (currentChannel, usersCertificateMapping) => {
+    if (!currentChannel) return []
+    const messagesArray = Object.values(currentChannel.messages)
+    return messagesArray.map(
+      message => {
+        if (usersCertificateMapping[message.pubKey] && message.id !== '44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a') {
```

Comment:
What is the deal with this hardcoded id?

Reply:
ouu, because I sent many of the first messages with the same id key :/

Reply:
There must be better way to solve this